### PR TITLE
Update documentation

### PR DESCRIPTION
**Pull request by phi-dbq**
## What changes are proposed in this pull request?

This PR updates README.md.
1. Create TOC and other style updates for README.md
2. Update documentation for Keras UDF

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,71 +1,83 @@
-# Deep Learning Pipelines for Apache Spark
+Deep Learning Pipelines for Apache Spark
+=============================================
+[![Build Status](https://travis-ci.org/databricks/spark-deep-learning.svg?branch=master)](https://travis-ci.org/databricks/spark-deep-learning)
 
-Deep Learning Pipelines provides high-level APIs for scalable deep learning in Python. The
```
Comment (thunterdb) [inline README.md]: Keep a high-level, 1- or 2-sentence description at the top. Readers must understand immediately what this project is about.

File: README.md [hunk 2]
```diff
@@ -1,71 +1,83 @@
...
-library comes from Databricks and leverages Spark for its two strongest facets:
-1. In the spirit of Spark and Spark MLlib, it provides easy-to-use APIs that enable deep learning
-in very few lines of code.
-2. It uses Spark's powerful distributed engine to scale out deep learning on massive datasets.
+- [Content](#content)
```
Comment (thunterdb) [inline README.md]: Do not put `Content` as a top level, it is a single bullet.

File: README.md [hunk 3]
```diff
@@ -1,71 +1,83 @@
...
-Currently, TensorFlow and TensorFlow-backed Keras workflows are supported, with a focus on model
-application and transfer learning on image data at scale, with hyper-parameter tuning in the works.
-Furthermore, it provides tools for data scientists and machine learning experts to turn deep
-learning models into SQL functions that can be used by a much wider group of users. It does not
-perform single-model distributed training - this is an area of active research, and here we aim to
-provide the most practical solutions for the majority of deep learning use cases.
 
-For an overview of the library, see the Databrick [blog post](https://databricks.com/blog/2017/06/06/databricks-vision-simplify-large-scale-deep-learning.html?preview=true) introducing Deep Learning Pipelines.
-For the various use cases the package serves, see the [Quick user guide](#quick-user-guide) section below.
+<a id="content"></a>
```
Comment (thunterdb) [inline README.md]: I thought it was automatically generated from the markdown. Try without it?

File: README.md [hunk 4]
```diff
@@ -1,71 +1,83 @@
...
+
+# Content
```
Comment (thunterdb) [inline README.md]: Overview

File: README.md [hunk 5]
```diff
@@ -1,71 +1,83 @@
...
+
+1.  In the spirit of Spark and Spark MLlib, it provides easy-to-use APIs that enable deep learning in very few lines of code.
+2.  It uses Spark's powerful distributed engine to scale out deep learning on massive datasets.
+
+Currently, TensorFlow and TensorFlow-backed Keras workflows are supported, with a focus on model application and transfer learning on image data at scale, with hyper-parameter tuning in the works. Furthermore, it provides tools for data scientists and machine learning experts to turn deep learning models into SQL functions that can be used by a much wider group of users. It does not perform single-model distributed training - this is an area of active research, and here we aim to provide the most practical solutions for the majority of deep learning use cases.
+
+For an overview of the library, see the Databrick [blog post](https://databricks.com/blog/2017/06/06/databricks-vision-simplify-large-scale-deep-learning.html?preview=true) introducing Deep Learning Pipelines. For the various use cases the package serves, see the [Quick user guide](#quick-user-guide) section below.
```
Comment (thunterdb) [inline README.md]: `Databricks`

File: README.md [hunk 6]
```diff
@@ -100,105 +112,134 @@ print("Training set accuracy = " + str(evaluator.evaluate(predictionAndLabels)))
...
+    transformer = TFImageTransformer(inputCol="image", outputCol="predictions", graph=frozen_graph,
+                                     inputTensor=image_arr, outputTensor=resized_images,
+                                     outputMode="image")
+    image_df = readImages("/data/myimages")
+    processed_image_df = transformer.transform(image_df)
+    ```
+
+3.  For Keras users
+
+    For applying Keras models in a distributed manner using Spark, [`KerasImageFileTransformer`](link_here) works on TensorFlow-backed Keras models. It \* Internally creates a DataFrame containing a column of images by applying the user-specified image loading and processing function to the input DataFrame containing a column of image URIs \* Loads a Keras model from the given model file path \* Applies the model to the image DataFrame
```
Comment (thunterdb) [inline README.md]: link?

Comment (thunterdb) [inline README.md]: The list failed here

File: README.md [hunk 7]
```diff
@@ -100,105 +112,134 @@ print("Training set accuracy = " + str(evaluator.evaluate(predictionAndLabels)))
...
-final_df = transformer.transform(uri_df)
 ```
 
 
+<a id="releases"></a>
+
 ## Releases:
 
-**TBA**
+**TBA**
```
Comment (sueann) [inline README.md]: Might as well update with our first release info for 0.1.0.

We could also release 0.1.1 (0.2.0?) now that the UDF creation function is in.

File: README.md (hunk 6 above)
Comment (phi-dbq) [inline README.md]: Oh, it should have been just a highlight.

File: README.md (hunk 3 above)
Comment (phi-dbq) [inline README.md]: I was porting this from org-mode, but now I found out that Emacs has a pretty good markdown mode, so I switched to directly updating the .md file.

File: README.md [hunk 8]
```diff
@@ -1,34 +1,51 @@
...
-For the various use cases the package serves, see the [Quick user guide](#quick-user-guide) section below.
+- [Overview](#overview)
+- [Building and running unit tests](#building-and-running-unit-tests)
+- [Spark version compatibility](#spark-version-compatibility)
+- [Quick user guide](#quick-user-guide)
+  - [Working with images in Spark](#working-with-images-in-spark)
+  - [Transfer learning](#transfer-learning)
+  - [Applying deep learning models at scale](#applying-deep-learning-models-at-scale)
+  - [Deploying models as SQL functions](#deploying-models-as-sql-functions)
+- [Releases:](#releases)
```
Comment (sueann) [inline README.md]: nit: remove `:`

File: README.md [hunk 9]
```diff
@@ -1,34 +1,51 @@
...
+Currently, TensorFlow and TensorFlow-backed Keras workflows are supported, with a focus on model inference/scoring and transfer learning on image data at scale, with hyper-parameter tuning in the works.
+
+Furthermore, it provides tools for data scientists and machine learning experts to turn deep learning models into SQL functions that can be used by a much wider group of users. It does not perform single-model distributed training - this is an area of active research, and here we aim to provide the most practical solutions for the majority of deep learning use cases.
+
+For an overview of the library, see the Databricks [blog post](https://databricks.com/blog/2017/06/06/databricks-vision-simplify-large-scale-deep-learning.html?preview=true) introducing Deep Learning Pipelines. For the various use cases the package serves, see the [Quick user guide](#quick-user-guide) section below.
 
 The library is in its early days, and we welcome everyone's feedback and contribution.
 
-Authors: Bago Amirbekian, Joseph Bradley, Sue Ann Hong, Tim Hunter, Philip Yang 
+Authors: Bago Amirbekian, Joseph Bradley, Sue Ann Hong, Tim Hunter, Philip Yang
```
Comment (sueann) [inline README.md]: Let's make this "Maintainers" instead of "Authors". Saw that somewhere and it felt nicer.

File: README.md [hunk 10]
```diff
@@ -1,34 +1,51 @@
...
-To compile this project, run `build/sbt assembly` from the project home directory.
-This will also run the Scala unit tests.
+To compile this project, run `build/sbt assembly` from the project home directory. This will also run the Scala unit tests.
+
+To run the Python unit tests, run the `run-tests.sh` script from the `python/` directory. You will need to set a few environment variables, e.g.
 
-To run the Python unit tests, run the `run-tests.sh` script from the `python/` directory.
-You will need to set a few environment variables, e.g.
-```bash
+```sh
```
Comment (sueann) [inline README.md]: can you add: "Be sure to run `build/sbt assembly` before running the Python tests." ?

File: README.md [hunk 11]
```diff
@@ -99,106 +107,132 @@ evaluator = MulticlassClassificationEvaluator(metricName="accuracy")
...
+
+    files = [os.path.abspath(os.path.join(dirpath, f)) for f in os.listdir("/data/myimages") if f.endswith('.jpg')]
+    uri_df = sqlContext.createDataFrame(files, StringType()).toDF("uri")
+
+    final_df = transformer.transform(uri_df)
+    ```
+
+
+### Deploying models as SQL functions
+
```
Comment (sueann) [inline README.md]: Let's add some more context:

"One way to productionize a model is to deploy it as a Spark SQL function, which allows anyone who knows SQL to use it. Deep Learning Pipelines provides mechanisms to take a deep learning model and register a Spark SQL user-defined function (UDF). "

File: README.md [hunk 12]
```diff
@@ -99,106 +107,132 @@ evaluator = MulticlassClassificationEvaluator(metricName="accuracy")
...
+The UDF takes a column a (formatted as a image struct "`SpImage`") and produces the output of the given Keras model (e.g. for [Inception V3](https://keras.io/applications/#inceptionv3), it produces a real valued score vector over the ImageNet object categories). For other models, the output could have different meanings. Please consult the actual models specification.
```
Comment (sueann) [inline README.md]: "The resulting UDF ..."

File: README.md [hunk 13]
```diff
@@ -99,106 +107,132 @@ evaluator = MulticlassClassificationEvaluator(metricName="accuracy")
...
-files = [os.path.abspath(os.path.join(dirpath, f)) for f in os.listdir("/data/myimages") if f.endswith('.jpg')]
-uri_df = sqlContext.createDataFrame(files, StringType()).toDF("uri")
-
-final_df = transformer.transform(uri_df)
+# Assume we have a compiled and trained Keras model
+model.save('path/to/my/model.h5')
+registerKerasImageUDF("my_custom_keras_model_udf", "path/to/my/model.h5")
+```
+
+If there are further preprocessing steps are required to prepare the images, the user has the option to provide a preprocessing function `preprocessor`. The `preprocessor` converts a file path into a image array. This function is usually introduced in Keras workflow, as in the following example.
```
Comment (sueann) [inline README.md]: The second "are" is not needed

File: README.md [hunk 14]
```diff
@@ -99,106 +107,132 @@ evaluator = MulticlassClassificationEvaluator(metricName="accuracy")
...
+
+def keras_load_img(fpath):
+    from keras.preprocessing.image import load_img, img_to_array
+    import numpy as np
+    img = load_img(fpath, target_size=(299, 299))
+    return img_to_array(img).astype(np.uint8)
+
+registerKerasImageUDF("my_keras_inception_udf", InceptionV3(weights="imagenet"), keras_load_img)
+
 ```
```
Comment (sueann) [inline README.md]: Add an example of what the SQL user's query would be like.

File: README.md [hunk 15]
```diff
@@ -196,7 +197,9 @@ Spark DataFrames are a natural construct for applying deep learning models to a
 
 ### Deploying models as SQL functions
 
-The UDF takes a column a (formatted as a image struct "`SpImage`") and produces the output of the given Keras model (e.g. for [Inception V3](https://keras.io/applications/#inceptionv3), it produces a real valued score vector over the ImageNet object categories). For other models, the output could have different meanings. Please consult the actual models specification.
+One way to productionize a model is to deploy it as a [Spark SQL User Defined Function](https://docs.databricks.com/spark/latest/spark-sql/udf-in-python.html), which allows anyone who knows SQL to use it. Deep Learning Pipelines provides mechanisms to take a deep learning model and register a Spark SQL User Defined Function (UDF).
+
+The resulting UDF takes a column a (formatted as a image struct "`SpImage`") and produces the output of the given Keras model (e.g. for [Inception V3](https://keras.io/applications/#inceptionv3), it produces a real valued score vector over the ImageNet object categories). For other models, the output could have different meanings. Please consult the actual models specification.
```
Comment (sueann) [inline README.md]: nit: "a column a" -> "a column"
